// @pixelium/web-vue/es/grid/index.js
import './css.js'
import { defineComponent, computed, ref, provide, createElementBlock, openBlock, normalizeStyle, renderSlot } from "vue";
import { u as useScreenWidth } from "../share/hook/use-screen-width.js";
import { k, G as Gn } from "../aside/index.js";
import { e as emitParentUpdate } from "../share/hook/use-index-of-children.js";
import { G as GRID_UPDATE } from "../share/const/event-bus-key.js";
import { G as GRID_PROVIDE } from "../share/const/provide-key.js";
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "Grid"
  },
  __name: "index",
  props: {
    column: { default: 24 },
    gutter: { default: 0 }
  },
  setup(__props) {
    const props = __props;
    const [widthType] = useScreenWidth();
    const gutterComputed = computed(() => {
      if (k(props.gutter)) {
        return {
          x: props.gutter,
          y: props.gutter
        };
      }
      if ("x" in props.gutter || "y" in props.gutter) {
        return {
          x: props.gutter.x || 0,
          y: props.gutter.y || 0
        };
      }
      const currentGutter = props.gutter[widthType.value];
      if (k(currentGutter)) {
        return {
          x: currentGutter,
          y: currentGutter
        };
      }
      return Gn({ x: 0, y: 0 }, currentGutter);
    });
    const columnComputed = computed(() => {
      if (k(props.column)) {
        return props.column;
      }
      const currentColumn = props.column[widthType.value];
      return currentColumn || 24;
    });
    const itemsStat = ref([]);
    provide(GRID_PROVIDE, {
      column: columnComputed,
      gutter: gutterComputed,
      itemsStat
    });
    emitParentUpdate(GRID_UPDATE);
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock(
        "div",
        {
          class: "pixelium px-grid",
          style: normalizeStyle({
            gridTemplateColumns: `repeat(${columnComputed.value}, minmax(0, 1fr))`,
            gap: gutterComputed.value ? `${gutterComputed.value.y}px ${gutterComputed.value.x}px` : void 0
          })
        },
        [
          renderSlot(_ctx.$slots, "default")
        ],
        4
        /* STYLE */
      );
    };
  }
});
export {
  _sfc_main as _
};
