// @pixelium/web-vue/es/index.js
import { _ as _sfc_main } from "./button/index.js";
import { _ as _sfc_main$1 } from "./button-group/index.js";
import { _ as _sfc_main$2 } from "./link/index.js";
import { _ as _sfc_main$3 } from "./message-box/message-box-wrapped.js";
import "./message-box/message-box.js";
import { m as message } from "./message-box/index.js";
import { _ as _sfc_main$4 } from "./row/index.js";
import { _ as _sfc_main$5 } from "./col/index.js";
import { _ as _sfc_main$6 } from "./divider/index.js";
import { _ as _sfc_main$7 } from "./space/index.js";
import { _ as _sfc_main$8 } from "./grid/index.js";
import { _ as _sfc_main$9 } from "./grid-item/index.js";
import { _ as _sfc_main$a } from "./container/index.js";
import { _ as _sfc_main$b } from "./main/index.js";
import { _ as _sfc_main$c } from "./aside/index.js";
import { _ as _sfc_main$d } from "./header/index.js";
import { _ as _sfc_main$e } from "./footer/index.js";
import { s } from "./share/util/theme.js";
import { u as useThemeMode } from "./share/hook/use-theme-mode.js";
import { _ as _sfc_main$f } from "./icon/index.js";
import { i as inBrowser } from "./share/util/env.js";
import { _ as _sfc_main$g } from "./input/index.js";
import { _ as _sfc_main$h } from "./input-number/index.js";
import { _ as _sfc_main$i } from "./textarea/index.js";
import { _ as _sfc_main$l } from "./tag/index.js";
import { _ as _sfc_main$j } from "./input-group/index.js";
import { _ as _sfc_main$k } from "./input-group-label/index.js";
import { _ as _sfc_main$m } from "./input-tag/index.js";
import { _ as _sfc_main$n } from "./tooltip/index.js";
import { _ as _sfc_main$o } from "./popover/index.js";
import { _ as _sfc_main$p } from "./empty/index.js";
import { _ as _sfc_main$q } from "./auto-complete/index.js";
import { _ as _sfc_main$r } from "./mask/index.js";
import { _ as _sfc_main$s } from "./spin/index.js";
import { _ as _sfc_main$t } from "./select/index.js";
const components = [
  _sfc_main,
  _sfc_main$1,
  _sfc_main$2,
  _sfc_main$3,
  _sfc_main$4,
  _sfc_main$5,
  _sfc_main$6,
  _sfc_main$7,
  _sfc_main$8,
  _sfc_main$9,
  _sfc_main$a,
  _sfc_main$b,
  _sfc_main$c,
  _sfc_main$d,
  _sfc_main$e,
  _sfc_main$f,
  _sfc_main$g,
  _sfc_main$h,
  _sfc_main$i,
  _sfc_main$j,
  _sfc_main$k,
  _sfc_main$l,
  _sfc_main$m,
  _sfc_main$n,
  _sfc_main$o,
  _sfc_main$p,
  _sfc_main$q,
  _sfc_main$r,
  _sfc_main$s,
  _sfc_main$t
];
const defaultPrefix = "Px";
const install = (app, options = { prefix: defaultPrefix }) => {
  components.forEach((component) => {
    if (component.name) {
      app.component(options.prefix + component.name, component);
    }
  });
  if (options.attachToApp !== false) {
    app.config.globalProperties.PixeliumVue = {
      message,
      useThemeMode
    };
  }
  if (options.attachToWindow !== false && inBrowser()) {
    window.$message = message;
  }
};
const index = {
  install
};
export {
  _sfc_main$c as Aside,
  _sfc_main$q as AutoComplete,
  _sfc_main as Button,
  _sfc_main$1 as ButtonGroup,
  _sfc_main$5 as Col,
  _sfc_main$a as Container,
  _sfc_main$6 as Divider,
  _sfc_main$p as Empty,
  _sfc_main$e as Footer,
  _sfc_main$8 as Grid,
  _sfc_main$9 as GridItem,
  _sfc_main$d as Header,
  _sfc_main$f as Icon,
  _sfc_main$g as Input,
  _sfc_main$j as InputGroup,
  _sfc_main$k as InputGroupLabel,
  _sfc_main$h as InputNumber,
  _sfc_main$m as InputTag,
  _sfc_main$2 as Link,
  _sfc_main$b as Main,
  _sfc_main$r as Mask,
  message as Message,
  _sfc_main$3 as MessageBox,
  _sfc_main$o as Popover,
  _sfc_main$4 as Row,
  _sfc_main$t as Select,
  _sfc_main$7 as Space,
  _sfc_main$s as Spin,
  _sfc_main$l as Tag,
  _sfc_main$i as Textarea,
  _sfc_main$n as Tooltip,
  index as default,
  install,
  s as setThemeColor,
  useThemeMode
};
