// @pixelium/web-vue/es/header/index.js
import './css.js'
import { defineComponent, computed, createElementBlock, openBlock, normalizeStyle, normalizeClass, unref, renderSlot } from "vue";
import { N, k } from "../aside/index.js";
import { u as useDarkMode } from "../share/hook/use-dark-mode.js";
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{ name: "Header" },
  __name: "index",
  props: {
    minHeight: null,
    bordered: { type: Boolean },
    dark: { type: Boolean, default: false }
  },
  setup(__props) {
    const props = __props;
    const darkMode = useDarkMode();
    const minHeight = computed(() => {
      return N(props.minHeight) ? void 0 : k(props.minHeight) ? `${props.minHeight}px` : props.minHeight;
    });
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock(
        "header",
        {
          class: normalizeClass(["pixelium px-header", {
            "px-header__dark": !unref(darkMode) && !!props.dark,
            "px-header__bordered": !!props.bordered
          }]),
          style: normalizeStyle({
            minHeight: minHeight.value
          })
        },
        [
          renderSlot(_ctx.$slots, "default")
        ],
        6
        /* CLASS, STYLE */
      );
    };
  }
});
export {
  _sfc_main as _
};
