// @pixelium/web-vue/es/option-list/index.js
import './css.js'
import { defineComponent, createElementBlock, openBlock, Fragment, renderList, unref, normalizeClass, renderSlot, createTextVNode, toDisplayString, createElementVNode } from "vue";
import { p, d } from "../aside/index.js";
import { G as GROUP_OPTION_TYPE } from "../share/const/index.js";
const _hoisted_1 = { class: "px-option-list" };
const _hoisted_2 = ["onClick"];
const _hoisted_3 = ["onClick"];
const _hoisted_4 = {
  key: 2,
  class: "px-option-list-item-group"
};
const _hoisted_5 = { class: "px-option-list-item-group-label" };
const _hoisted_6 = ["onClick"];
const _hoisted_7 = ["onClick"];
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "OptionList"
  },
  __name: "index",
  props: {
    options: { default: () => [] },
    activeValues: { default: () => [] }
  },
  emits: ["select"],
  setup(__props, { emit: __emit }) {
    const props = __props;
    const emits = __emit;
    const selectHandler = (value, e) => {
      emits("select", value, value, e);
    };
    const selectObjectHandler = (option, e) => {
      if (option.disabled) {
        return;
      }
      emits("select", option.value, option, e);
    };
    const getKey = (option) => {
      if (p(option)) {
        return option;
      } else if ("type" in option && option.type === GROUP_OPTION_TYPE) {
        return option.key;
      } else {
        return option.label;
      }
    };
    const checkActive = (option) => {
      if (!props.activeValues.length) {
        return false;
      }
      if (p(option)) {
        return props.activeValues.includes(option);
      }
      return props.activeValues.includes(option.value);
    };
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("ul", _hoisted_1, [
        (openBlock(true), createElementBlock(
          Fragment,
          null,
          renderList(props.options, (item) => {
            return openBlock(), createElementBlock(
              Fragment,
              {
                key: getKey(item)
              },
              [
                unref(p)(item) ? (openBlock(), createElementBlock("li", {
                  key: 0,
                  class: normalizeClass(["px-option-list-item", {
                    "px-option-list-item__active": checkActive(item)
                  }]),
                  onClick: (e) => selectHandler(item, e)
                }, [
                  renderSlot(_ctx.$slots, "option", { option: item }, () => [
                    createTextVNode(
                      toDisplayString(item),
                      1
                      /* TEXT */
                    )
                  ])
                ], 10, _hoisted_2)) : unref(d)(item) && !("type" in item && item.type === unref(GROUP_OPTION_TYPE)) ? (openBlock(), createElementBlock("li", {
                  key: 1,
                  class: normalizeClass(["px-option-list-item", {
                    "px-option-list-item__disabled": item.disabled,
                    "px-option-list-item__active": checkActive(item)
                  }]),
                  onClick: (e) => selectObjectHandler(item, e)
                }, [
                  renderSlot(_ctx.$slots, "option", { option: item }, () => [
                    createTextVNode(
                      toDisplayString(item.label),
                      1
                      /* TEXT */
                    )
                  ])
                ], 10, _hoisted_3)) : (openBlock(), createElementBlock("li", _hoisted_4, [
                  createElementVNode("div", _hoisted_5, [
                    renderSlot(_ctx.$slots, "group-label", { option: item }, () => [
                      createTextVNode(
                        toDisplayString(item.label),
                        1
                        /* TEXT */
                      )
                    ])
                  ]),
                  (openBlock(true), createElementBlock(
                    Fragment,
                    null,
                    renderList(item.children, (child) => {
                      return openBlock(), createElementBlock(
                        Fragment,
                        {
                          key: getKey(child)
                        },
                        [
                          unref(p)(child) ? (openBlock(), createElementBlock("li", {
                            key: 0,
                            class: normalizeClass(["px-option-list-item", {
                              "px-option-list-item__active": checkActive(child)
                            }]),
                            onClick: (e) => selectHandler(child, e)
                          }, [
                            renderSlot(_ctx.$slots, "option", { option: child }, () => [
                              createTextVNode(
                                toDisplayString(child),
                                1
                                /* TEXT */
                              )
                            ])
                          ], 10, _hoisted_6)) : (openBlock(), createElementBlock("li", {
                            key: 1,
                            class: normalizeClass(["px-option-list-item", {
                              "px-option-list-item__disabled": child.disabled,
                              "px-option-list-item__active": checkActive(child)
                            }]),
                            onClick: (e) => selectObjectHandler(child, e)
                          }, [
                            renderSlot(_ctx.$slots, "option", { option: child }, () => [
                              createTextVNode(
                                toDisplayString(child.label),
                                1
                                /* TEXT */
                              )
                            ])
                          ], 10, _hoisted_7))
                        ],
                        64
                        /* STABLE_FRAGMENT */
                      );
                    }),
                    128
                    /* KEYED_FRAGMENT */
                  ))
                ]))
              ],
              64
              /* STABLE_FRAGMENT */
            );
          }),
          128
          /* KEYED_FRAGMENT */
        ))
      ]);
    };
  }
});
export {
  _sfc_main as _
};
