// @pixelium/web-vue/es/empty/index.js
import './css.js'
import { defineComponent, createElementBlock, openBlock, normalizeStyle, createElementVNode, renderSlot, createVNode, unref, createTextVNode, toDisplayString } from "vue";
const FileDelete = /* @__PURE__ */ defineComponent({
  props: {
    size: {},
    color: {}
  },
  setup(__props) {
    const props = __props;
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock(
        "svg",
        {
          class: "px-icon-pa",
          style: normalizeStyle({ color: props.color, fontSize: props.size + "px" }),
          xmlns: "http://www.w3.org/2000/svg",
          fill: "none",
          viewBox: "0 0 24 24"
        },
        [..._cache[0] || (_cache[0] = [
          createElementVNode(
            "path",
            {
              fill: "currentColor",
              d: "M11 22h10V8h-2V6h-2v2h-2V6h2V4h-2V2H3v12h2V4h8v6h6v10h-8zm-4-2H5v2H3v-2h2v-2H3v-2h2v2h2v-2h2v2H7zm0 0h2v2H7z"
            },
            null,
            -1
            /* CACHED */
          )
        ])],
        4
        /* STYLE */
      );
    };
  }
});
const _hoisted_1 = { class: "pixelium px-empty" };
const _hoisted_2 = { class: "px-empty-icon-wrapper" };
const _hoisted_3 = { class: "px-empty-description" };
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "Empty"
  },
  __name: "index",
  props: {
    description: { default: "No Data" }
  },
  setup(__props) {
    const props = __props;
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", _hoisted_1, [
        createElementVNode("div", _hoisted_2, [
          renderSlot(_ctx.$slots, "image", {}, () => [
            createVNode(unref(FileDelete), { class: "px-empty-icon" })
          ])
        ]),
        createElementVNode("div", _hoisted_3, [
          renderSlot(_ctx.$slots, "description", {}, () => [
            createTextVNode(
              toDisplayString(props.description),
              1
              /* TEXT */
            )
          ])
        ])
      ]);
    };
  }
});
export {
  _sfc_main as _
};
