// @pixelium/web-vue/es/row/index.js
import './css.js'
import { defineComponent, computed, provide, createElementBlock, openBlock, createElementVNode, normalizeStyle, normalizeClass, renderSlot } from "vue";
import { k, G as Gn } from "../aside/index.js";
import { u as useScreenWidth } from "../share/hook/use-screen-width.js";
const _hoisted_1 = { class: "pixelium px-row" };
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "Row"
  },
  __name: "index",
  props: {
    gutter: { default: 0 },
    justify: { default: "start" },
    align: { default: "start" },
    wrap: { type: Boolean, default: true }
  },
  setup(__props) {
    const props = __props;
    const [widthType] = useScreenWidth();
    const gutterComputed = computed(() => {
      if (k(props.gutter)) {
        return {
          x: props.gutter,
          y: props.gutter
        };
      }
      if ("x" in props.gutter || "y" in props.gutter) {
        return {
          x: props.gutter.x || 0,
          y: props.gutter.y || 0
        };
      }
      const currentGutter = props.gutter[widthType.value];
      if (k(currentGutter)) {
        return {
          x: currentGutter,
          y: currentGutter
        };
      }
      return Gn({ x: 0, y: 0 }, currentGutter);
    });
    provide("px-row-provide", gutterComputed);
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", _hoisted_1, [
        createElementVNode(
          "div",
          {
            class: normalizeClass(["px-row-inner", {
              [`px-row__justify-${props.justify}`]: true,
              [`px-row__align-${props.align}`]: true,
              [`px-row__wrap`]: !!props.wrap
            }]),
            style: normalizeStyle({
              marginLeft: gutterComputed.value ? -gutterComputed.value.x / 2 + "px" : void 0,
              marginRight: gutterComputed.value ? -gutterComputed.value.x / 2 + "px" : void 0,
              marginTop: gutterComputed.value ? -gutterComputed.value.y / 2 + "px" : void 0,
              marginBottom: gutterComputed.value ? -gutterComputed.value.y / 2 + "px" : void 0
            })
          },
          [
            renderSlot(_ctx.$slots, "default")
          ],
          6
          /* CLASS, STYLE */
        )
      ]);
    };
  }
});
export {
  _sfc_main as _
};
