// @pixelium/web-vue/es/button-group/index.js
import './css.js'
import { defineComponent, provide, createElementBlock, openBlock, renderSlot } from "vue";
import { e as emitParentUpdate } from "../share/hook/use-index-of-children.js";
import { B as BUTTON_GROUP_UPDATE } from "../share/const/event-bus-key.js";
import { B as BUTTON_GROUP_PROVIDE } from "../share/const/provide-key.js";
const _hoisted_1 = { class: "pixelium px-button-group" };
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "ButtonGroup"
  },
  __name: "index",
  props: {
    borderRadius: null,
    shape: { default: "default" },
    size: { default: "medium" },
    disabled: { type: Boolean, default: false },
    loading: { type: Boolean },
    variant: null
  },
  setup(__props) {
    const props = __props;
    provide(BUTTON_GROUP_PROVIDE, props);
    emitParentUpdate(BUTTON_GROUP_UPDATE);
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", _hoisted_1, [
        renderSlot(_ctx.$slots, "default")
      ]);
    };
  }
});
export {
  _sfc_main as _
};
