// @pixelium/web-vue/es/input-group/index.js
import './css.js'
import { defineComponent, provide, createElementBlock, openBlock, renderSlot } from "vue";
import { e as emitParentUpdate } from "../share/hook/use-index-of-children.js";
import { I as INPUT_GROUP_UPDATE } from "../share/const/event-bus-key.js";
import { I as INPUT_GROUP_PROVIDE } from "../share/const/provide-key.js";
const _hoisted_1 = {
  class: "pixelium px-input-group",
  ref: "slotWrapper"
};
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "InputGroup"
  },
  __name: "index",
  props: {
    borderRadius: null,
    shape: { default: "default" },
    size: { default: "medium" },
    disabled: { type: Boolean, default: false }
  },
  setup(__props) {
    const props = __props;
    provide(INPUT_GROUP_PROVIDE, props);
    emitParentUpdate(INPUT_GROUP_UPDATE);
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock(
        "div",
        _hoisted_1,
        [
          renderSlot(_ctx.$slots, "default")
        ],
        512
        /* NEED_PATCH */
      );
    };
  }
});
export {
  _sfc_main as _
};
