// @pixelium/web-vue/es/aside/index.js
import './css.js'
import { defineComponent, getCurrentInstance, ref, inject, onMounted, onBeforeUnmount, computed, createElementBlock, openBlock, normalizeStyle, normalizeClass, unref, renderSlot } from "vue";
import { _ as _sfc_main$1 } from "../container/index.js";
import { u as useDarkMode } from "../share/hook/use-dark-mode.js";
const Q = /\.|\[(\d+)\]/;
function M(n) {
  return n.split(Q).filter(Boolean);
}
function U(n) {
  return n.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
function d(n) {
  return n !== null && (typeof n == "object" || typeof n == "function");
}
function f(n, e, t) {
  const s = M(e), o = s.length;
  let c = n;
  for (let i = 0; i < o; i++)
    if (d(c))
      c = c[s[i]];
    else
      return t;
  return t ?? c;
}
function m(n) {
  return typeof n == "function";
}
function Cn(n, e, t = {}) {
  const { start: s = "{", end: o = "}" } = t, c = new RegExp(
    `${U(s)}(.*?)${U(o)}`,
    "g"
  );
  return n.replace(c, (i, r) => {
    const l = m(e) ? e(r) : f(e, r);
    return l ?? i;
  });
}
function p(n) {
  return typeof n == "string" || n instanceof String;
}
function Pn(n) {
  return n === null;
}
function k(n) {
  return typeof n == "number" || n instanceof Number;
}
function en(n) {
  return n === void 0;
}
function $(n) {
  return Array.isArray(n);
}
function R(n) {
  return k(n) && (typeof n == "number" ? n === 1 / 0 || n === -1 / 0 : n.valueOf() === 1 / 0 || n.valueOf() === -1 / 0);
}
function z(n) {
  return k(n) && (typeof n == "number" ? Number.isNaN(n) : Number.isNaN(n.valueOf()));
}
function N(n) {
  return n == null;
}
function Gn(...n) {
  return n.reduce((e, t) => {
    if (N(t))
      return e;
    const s = Object.keys(t);
    for (const o of s) {
      const c = e[o], i = t[o];
      (N(c) || !N(i)) && (e[o] = i);
    }
    return e;
  }, {});
}
function Ne(n, e, t) {
  let s, o;
  const { immediate: c = false, maxWait: i } = t || {};
  return function(...l) {
    i !== void 0 && !o && (o = setTimeout(() => {
      o = void 0, n.apply(this, l);
    }, i)), !s && c && n.apply(this, l), s && clearTimeout(s), s = setTimeout(() => {
      o && (clearTimeout(o), o = void 0), n.apply(this, l);
    }, e);
  };
}
// @ts-ignore
typeof process < "u" && process.env.NODE_ENV !== "production";
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{ name: "Aside" },
  __name: "index",
  props: {
    width: null,
    bordered: { type: Boolean, default: false },
    side: { default: "left" },
    dark: { type: Boolean, default: false }
  },
  setup(__props) {
    var _a;
    const instance = getCurrentInstance();
    const inner = ref(((_a = instance == null ? void 0 : instance.parent) == null ? void 0 : _a.type) === _sfc_main$1);
    const asideCounter = inner.value ? inject("px-container-provide") : void 0;
    onMounted(() => {
      if (asideCounter) {
        asideCounter.value += 1;
      }
    });
    onBeforeUnmount(() => {
      if (asideCounter) {
        asideCounter.value -= 1;
      }
    });
    const props = __props;
    const darkMode = useDarkMode();
    const width = computed(() => {
      return N(props.width) ? void 0 : k(props.width) ? `${props.width}px` : props.width;
    });
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock(
        "aside",
        {
          class: normalizeClass(["pixelium px-aside", {
            "px-aside__dark": !unref(darkMode) && !!props.dark,
            "px-aside__bordered": !!props.bordered,
            [`px-aside__${props.side}`]: true
          }]),
          style: normalizeStyle({
            width: width.value
          })
        },
        [
          renderSlot(_ctx.$slots, "default")
        ],
        6
        /* CLASS, STYLE */
      );
    };
  }
});
export {
  $,
  Cn as C,
  Gn as G,
  N,
  Pn as P,
  R,
  _sfc_main as _,
  Ne as a,
  d,
  en as e,
  k,
  m,
  p,
  z
};
