// @pixelium/web-vue/es/spin/index.js
import './css.js'
import { defineComponent, useSlots, computed, createElementBlock, openBlock, normalizeStyle, normalizeClass, unref, renderSlot, createCommentVNode, createBlock, createElementVNode, createVNode } from "vue";
import { C as COVER_Z_INDEX } from "../share/const/index.js";
import { _ as _sfc_main$1 } from "../mask/index.js";
import { S as SpinnerThirdSolid } from "../auto-complete/index.js";
import { k, p } from "../aside/index.js";
const _hoisted_1 = { class: "px-spin-content" };
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "Spin"
  },
  __name: "index",
  props: {
    loading: { type: Boolean, default: void 0 },
    size: { default: "medium" },
    maskColor: null,
    maskStep: { default: 1 },
    maskLineWidth: { default: 2 },
    maskGrid: { type: Boolean, default: true },
    zIndex: { default: COVER_Z_INDEX }
  },
  setup(__props) {
    const props = __props;
    const slots = useSlots();
    const sizeValue = computed(() => {
      return k(props.size) ? `${props.size}px` : void 0;
    });
    const loadingComputed = computed(() => {
      return slots.default ? !!props.loading : props.loading !== false;
    });
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock(
        "div",
        {
          class: normalizeClass(["pixelium px-spin", {
            [`px-spin__${props.size}`]: unref(p)(props.size)
          }]),
          style: normalizeStyle({
            minWidth: sizeValue.value,
            minHeight: sizeValue.value
          })
        },
        [
          renderSlot(_ctx.$slots, "default"),
          loadingComputed.value ? (openBlock(), createElementBlock(
            "div",
            {
              key: 0,
              class: "px-spin-cover",
              style: normalizeStyle({
                zIndex: props.zIndex
              })
            },
            [
              unref(slots).default ? (openBlock(), createBlock(_sfc_main$1, {
                key: 0,
                "z-index": 0,
                color: props.maskColor,
                step: props.maskStep,
                "line-width": props.maskLineWidth,
                grid: props.maskGrid
              }, null, 8, ["color", "step", "line-width", "grid"])) : createCommentVNode("v-if", true),
              createElementVNode("div", _hoisted_1, [
                createElementVNode(
                  "div",
                  {
                    class: normalizeClass(["px-spin-icon-wrapper", {
                      "px-spin-icon-wrapper__last": !unref(slots).description,
                      [`px-spin-icon-wrapper__${props.size}`]: unref(p)(props.size)
                    }]),
                    style: normalizeStyle({
                      fontSize: sizeValue.value
                    })
                  },
                  [
                    renderSlot(_ctx.$slots, "icon", {}, () => [
                      createVNode(unref(SpinnerThirdSolid), {
                        class: "px-spin-icon px-animation__loading",
                        style: normalizeStyle({
                          width: sizeValue.value,
                          height: sizeValue.value
                        })
                      }, null, 8, ["style"])
                    ])
                  ],
                  6
                  /* CLASS, STYLE */
                ),
                renderSlot(_ctx.$slots, "description")
              ])
            ],
            4
            /* STYLE */
          )) : createCommentVNode("v-if", true)
        ],
        6
        /* CLASS, STYLE */
      );
    };
  }
});
export {
  _sfc_main as _
};
