// @pixelium/web-vue/es/divider/index.js
import './css.js'
import { defineComponent, computed, createElementBlock, openBlock, normalizeStyle, normalizeClass } from "vue";
import { N, k } from "../aside/index.js";
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "Divider"
  },
  __name: "index",
  props: {
    direction: { default: "horizontal" },
    variant: { default: "solid" },
    margin: null,
    size: null
  },
  setup(__props) {
    const props = __props;
    const marginComputed = computed(() => {
      if (N(props.margin)) {
        return;
      }
      return k(props.margin) ? `${props.margin}px` : props.margin;
    });
    const sizeComputed = computed(() => {
      if (N(props.size)) {
        return;
      }
      return k(props.size) ? `${props.size}px` : props.size;
    });
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock(
        "div",
        {
          class: normalizeClass(["pixelium px-divider", {
            [`px-divider__${props.direction}`]: true,
            [`px-divider__${props.variant}`]: true
          }]),
          style: normalizeStyle({
            marginLeft: props.direction === "vertical" ? marginComputed.value : void 0,
            marginRight: props.direction === "vertical" ? marginComputed.value : void 0,
            marginTop: props.direction === "horizontal" ? marginComputed.value : void 0,
            marginBottom: props.direction === "horizontal" ? marginComputed.value : void 0,
            borderBottomWidth: props.direction === "horizontal" ? sizeComputed.value : void 0,
            borderRightWidth: props.direction === "vertical" ? sizeComputed.value : void 0
          })
        },
        null,
        6
        /* CLASS, STYLE */
      );
    };
  }
});
export {
  _sfc_main as _
};
