// @pixelium/web-vue/es/popup-trigger/index.js
import './css.js'
import { defineComponent, shallowRef, useSlots, createVNode, Fragment, cloneVNode, mergeProps } from "vue";
import { f as flattenVNodes, i as isTextVNode } from "../share/util/render.js";
import { u as useClickOutsideListener } from "../share/hook/use-click-outside-listener.js";
import { i as inBrowser } from "../share/util/env.js";
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "PopupTrigger"
  },
  __name: "index",
  props: {
    trigger: {
      default: "hover"
    },
    disabled: {
      type: Boolean
    }
  },
  emits: ["close", "open"],
  setup(__props, {
    expose: __expose,
    emit: __emit
  }) {
    const props = __props;
    const emits = __emit;
    const currentTrigger = shallowRef(null);
    async function openHandler(node, e) {
      if (inBrowser()) {
        if (node.el instanceof HTMLElement) {
          currentTrigger.value = node;
        } else {
          currentTrigger.value = null;
        }
      }
      if (props.disabled) {
        return;
      }
      emits("open", node, e);
    }
    async function closeHandler(e) {
      if (props.disabled) {
        return;
      }
      emits("close", e);
    }
    const slots = useSlots();
    const clickOutsideHandler = (event) => {
      if (props.trigger === "click") {
        closeHandler(event);
      }
    };
    useClickOutsideListener(currentTrigger, clickOutsideHandler);
    const firstVNodeRef = shallowRef(null);
    __expose({
      firstVNode: firstVNodeRef
    });
    return () => {
      var _a;
      const children = flattenVNodes(((_a = slots.default) == null ? void 0 : _a.call(slots)) || []);
      return createVNode(Fragment, null, [children.map((child, index) => {
        if (isTextVNode(child)) {
          const node = props.trigger === "click" ? createVNode("span", {
            "onClick": (e) => openHandler(node, e)
          }, [child]) : createVNode("span", {
            "onMouseenter": (e) => openHandler(node, e),
            "onMouseleave": (e) => closeHandler(e)
          }, [child]);
          if (index === 0) {
            firstVNodeRef.value = node;
            if (!currentTrigger.value) {
              currentTrigger.value = node;
            }
          }
          return node;
        } else {
          const clone = cloneVNode(child, mergeProps(child.props || {}, props.trigger === "click" ? {
            onClick: (e) => openHandler(clone, e)
          } : {
            onMouseenter: (e) => openHandler(clone, e),
            onMouseleave: (e) => closeHandler(e)
          }));
          clone.ref = child.ref;
          if (index === 0) {
            firstVNodeRef.value = clone;
            if (!currentTrigger.value) {
              currentTrigger.value = clone;
            }
          }
          return clone;
        }
      })]);
    };
  }
});
export {
  _sfc_main as _
};
