// parsnip-kit/dist/parsnip-kit.js
const Y = new RegExp("[-_\\s]|(?<=[a-z])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])|(\\d+)");
function A(n) {
  return n.split(Y).filter(Boolean);
}
function xn(n) {
  const e = A(n);
  let t = "";
  const s = e.length;
  for (let o = 0; o < s; o++)
    t += o > 0 ? e[o].charAt(0).toUpperCase() + e[o].slice(1).toLowerCase() : e[o].toLowerCase();
  return t;
}
function On(n) {
  const e = A(n);
  let t = "";
  const s = e.length;
  for (let o = 0; o < s; o++)
    t += e[o].toLowerCase() + (o < s - 1 ? "-" : "");
  return t;
}
function Mn(n) {
  const e = A(n);
  let t = "";
  const s = e.length;
  for (let o = 0; o < s; o++)
    t += e[o].charAt(0).toUpperCase() + e[o].slice(1).toLowerCase();
  return t;
}
function Tn(n) {
  const e = A(n);
  let t = "";
  const s = e.length;
  for (let o = 0; o < s; o++)
    t += e[o].toLowerCase() + (o < s - 1 ? "_" : "");
  return t;
}
const Q = /\.|\[(\d+)\]/;
function M(n) {
  return n.split(Q).filter(Boolean);
}
function In(n) {
  const e = A(n);
  let t = "";
  const s = e.length;
  for (let o = 0; o < s; o++)
    t += e[o].toUpperCase() + (o < s - 1 ? "_" : "");
  return t;
}
function X(n) {
  return n.charAt(0).toUpperCase() + n.slice(1).toLowerCase();
}
function Sn(n) {
  const e = A(n);
  let t = "";
  const s = e.length;
  for (let o = 0; o < s; o++)
    t += X(e[o]) + (o < s - 1 ? " " : "");
  return t;
}
const g = {
  "&": "&amp;",
  "<": "&lt;",
  ">": "&gt;",
  '"': "&quot;",
  "'": "&#39;"
};
function En(n) {
  let e = "";
  for (let t = 0; t < n.length; t++) {
    const s = n[t];
    e += g[s] || s;
  }
  return e;
}
const nn = {
  "&amp;": "&",
  "&lt;": "<",
  "&gt;": ">",
  "&quot;": '"',
  "&#34;": '"',
  "&#39;": "'"
};
function Bn(n) {
  return n.replace(/&lt;|&gt;|&amp;|&quot;|&#39;|&#34;/g, (e) => nn[e]);
}
function U(n) {
  return n.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
function d(n) {
  return n !== null && (typeof n == "object" || typeof n == "function");
}
function f(n, e, t) {
  const s = M(e), o = s.length;
  let c = n;
  for (let i = 0; i < o; i++)
    if (d(c))
      c = c[s[i]];
    else
      return t;
  return t ?? c;
}
function m(n) {
  return typeof n == "function";
}
function Cn(n, e, t = {}) {
  const { start: s = "{", end: o = "}" } = t, c = new RegExp(
    `${U(s)}(.*?)${U(o)}`,
    "g"
  );
  return n.replace(c, (i, r) => {
    const l = m(e) ? e(r) : f(e, r);
    return l ?? i;
  });
}
function p(n) {
  return typeof n == "string" || n instanceof String;
}
function Nn(n) {
  return typeof n == "boolean" || n instanceof Boolean;
}
function Pn(n) {
  return n === null;
}
function k(n) {
  return typeof n == "number" || n instanceof Number;
}
function en(n) {
  return n === void 0;
}
function $n(n) {
  return typeof n == "symbol";
}
function Fn(n) {
  return typeof n == "bigint";
}
function j(n) {
  return !d(n);
}
function v(n) {
  return p(n) && /^\d+$/.test(n);
}
function $(n) {
  return Array.isArray(n);
}
function tn(n) {
  return n !== null && typeof n == "object";
}
function Un(n) {
  return m(n) && d(n.prototype) && m(n.prototype.constructor);
}
function jn(n) {
  return d(n) && "length" in n && typeof n.length == "number";
}
function R(n) {
  return k(n) && (typeof n == "number" ? n === 1 / 0 || n === -1 / 0 : n.valueOf() === 1 / 0 || n.valueOf() === -1 / 0);
}
function z(n) {
  return k(n) && (typeof n == "number" ? Number.isNaN(n) : Number.isNaN(n.valueOf()));
}
function _n(n) {
  return k(n) && !R(n) && !z(n) && (typeof n == "number" ? n === Math.floor(n) : n.valueOf() === Math.floor(n));
}
function qn(n) {
  return k(n) && !R(n) && !z(n) && (typeof n == "number" ? n !== Math.floor(n) : n.valueOf() !== Math.floor(n));
}
const S = (n) => Object.prototype.toString.apply(n).slice(8, -1);
function N(n) {
  return n == null;
}
function on(n, e, t) {
  const s = M(e), o = s.length;
  let c = n;
  for (let i = 0; i < o; i++) {
    let r;
    if (v(s[i]) && (r = parseInt(s[i])), i === o - 1)
      c[r === void 0 ? s[i] : r] = t;
    else {
      const l = v(s[i + 1]), u = r === void 0 ? s[i] : r;
      d(c[u]) || (c[u] = l ? [] : {}), c = c[u];
    }
  }
}
function Ln(n, e) {
  const t = M(e), s = t.length;
  let o = n;
  for (let c = 0; c < s; c++)
    if (d(o))
      c < s - 1 ? o = o[t[c]] : v(t[c]) && Array.isArray(o) ? o.splice(parseInt(t[c]), 1) : delete o[t[c]];
    else
      throw TypeError(
        `${t.slice(0, c + 1).reduce(
          (i, r, l) => i + (v(r) ? `[${r}]` : l === 0 ? `${r}` : `.${r}`),
          ""
        )} is not an object.`
      );
}
function Dn(n) {
  if (tn(n)) {
    let e = {};
    const t = S(n);
    if (n instanceof Map) {
      e = /* @__PURE__ */ new Map();
      for (const s of n.entries())
        e.set(s[0], s[1]);
    } else if (n instanceof Set) {
      e = /* @__PURE__ */ new Set();
      for (const s of n.values())
        e.add(s);
    } else if (Array.isArray(n) || t === "Arguments") {
      e = new Array(n.length);
      const s = Object.keys(n);
      for (const o of s)
        e[o] = n[o];
    } else if (e = W(n, t), !e)
      if (t === "Object") {
        n.__proto__ ? (e = Object.create(n.__proto__), Object.defineProperty(e, "constructor", {
          value: n.constructor,
          writable: !0,
          enumerable: !1,
          configurable: !0
        })) : e = {};
        const s = Object.keys(n);
        for (const o of s)
          e[o] = n[o];
      } else
        e = {};
    return e;
  } else
    return n;
}
const W = (n, e, t) => {
  if (t && t.has(n))
    return t.get(n);
  let s;
  switch (e) {
    case "Number":
      s = new Number(n.valueOf());
      break;
    case "String":
      s = new String(n.valueOf());
      break;
    case "Boolean":
      s = new Boolean(n.valueOf());
      break;
    case "Date":
      s = new Date(n.valueOf());
      break;
    case "RegExp":
      s = new RegExp(n.source, n.flags);
      break;
    case "Blob":
      s = new Blob([n], { type: n.type });
      break;
    case "File":
      s = new File([n], n.name, { type: n.type });
      break;
    case "ArrayBuffer":
    case "Int8Array":
    case "Uint8Array":
    case "Uint8ClampedArray":
    case "Int16Array":
    case "Uint16Array":
    case "Int32Array":
    case "Uint32Array":
    case "Float32Array":
    case "Float64Array":
    case "BigInt64Array":
    case "BigUint64Array":
      s = n.slice();
      break;
  }
  return t && s && t.set(n, s), s;
};
function Rn(n, e) {
  return K(n, /* @__PURE__ */ new WeakMap(), void 0, e);
}
function x(n, e, t, s) {
  let o = n;
  if (d(n))
    if (e.has(n))
      o = e.get(n);
    else {
      const c = K(n, e, t, s);
      c && (o = c, e.set(n, c));
    }
  return o;
}
function K(n, e, t, s) {
  if (d(n)) {
    let o = {};
    const c = S(n);
    if (n instanceof Map) {
      o = /* @__PURE__ */ new Map();
      for (const i of n.entries())
        o.set(
          x(i[0], e, void 0, s),
          x(i[1], e, void 0, s)
        );
    } else if (n instanceof Set) {
      o = /* @__PURE__ */ new Set();
      for (const i of n.values())
        o.add(x(i, e, void 0, s));
    } else if (Array.isArray(n) || c === "Arguments") {
      o = new Array(n.length);
      const i = Object.keys(n);
      for (const r of i)
        o[r] = x(n[r], e, r, s);
    } else
      o = W(n, c, e), o || (m(s) ? o = s(n, t, e, _) : c === "Object" ? o = _(n, e, s) : o = n);
    return o;
  } else
    return n;
}
function _(n, e, t) {
  if (e.has(n))
    return e.get(n);
  let s;
  n.__proto__ ? (s = Object.create(n.__proto__), Object.defineProperty(s, "constructor", {
    value: n.constructor,
    writable: !0,
    enumerable: !1,
    configurable: !0
  })) : s = {}, e.set(n, s);
  const o = Object.keys(n);
  for (const c of o)
    s[c] = x(n[c], e, c, t);
  return s;
}
function zn(n, e) {
  const t = e.map((c) => M(c)[0]).map((c) => v(c) ? parseInt(c) : c), s = Array.isArray(n) ? [] : {}, o = Object.keys(n);
  for (const c of o)
    s[c] = n[c];
  return t.sort((c, i) => {
    const r = k(c), l = k(i);
    return r && l ? i - c : r || l ? r ? -1 : 1 : c.localeCompare(i);
  }), t.forEach((c) => {
    Array.isArray(s) && k(c) ? s.splice(c, 1) : delete s[c];
  }), s;
}
function Wn(n, e) {
  const t = e.map((o) => M(o)[0]), s = Array.isArray(n) ? [] : {};
  return t.forEach((o) => {
    Object.prototype.hasOwnProperty.call(n, o) && (Array.isArray(s) ? s.push(n[o]) : s[o] = n[o]);
  }), s;
}
function Kn(n, e) {
  const t = Array.isArray(n) ? [] : {}, s = Object.keys(n);
  for (const c of s)
    t[c] = n[c];
  let o = 0;
  for (const c of s)
    e(n[c], c, n) || (Array.isArray(t) && v(c) ? (t.splice(parseInt(c) - o, 1), o++) : delete t[c]);
  return t;
}
function Vn(n, e) {
  const t = Object.keys(n);
  for (const s of t)
    e(n[s], s, n);
}
function O(n, e) {
  return Object.is(n, e);
}
function E(n, e) {
  return sn(n, e);
}
function sn(n, e) {
  if (j(n) || j(e))
    return O(n, e);
  if (O(n, e))
    return !0;
  const t = S(n), s = S(e);
  if (t !== s)
    return !1;
  if (Array.isArray(n))
    return q(n, e);
  switch (t) {
    case "Number":
    case "Boolean":
    case "Date":
      return O(n.valueOf(), e.valueOf());
    case "String":
    case "RegExp":
      return n.toString() === e.toString();
    case "ArrayBuffer":
      return un(n, e);
    case "Int8Array":
    case "Uint8Array":
    case "Uint8ClampedArray":
    case "Int16Array":
    case "Uint16Array":
    case "Int32Array":
    case "Uint32Array":
    case "Float32Array":
    case "Float64Array":
    case "BigInt64Array":
    case "BigUint64Array":
      return V(n, e);
    case "Object":
      return q(n, e);
    case "Set":
      return rn(n, e);
    case "Map":
      return cn(n, e);
    default:
      return O(n, e);
  }
}
function cn(n, e) {
  if (n.size !== e.size)
    return !1;
  const t = [...n.keys()], s = [...e.keys()], o = n.size;
  for (let c = 0; c < o; c++)
    if (!E(t[c], s[c]) || !E(n.get(t[c]), e.get(s[c])))
      return !1;
  return !0;
}
function rn(n, e) {
  if (n.size !== e.size)
    return !1;
  const t = [...n.values()], s = [...e.values()], o = n.size;
  for (let c = 0; c < o; c++)
    if (!E(t[c], s[c]))
      return !1;
  return !0;
}
function q(n, e) {
  const t = Object.keys(n), s = Object.keys(e);
  if (t.length !== s.length)
    return !1;
  const o = t.length;
  for (let c = 0; c < o; c++)
    if (!Object.prototype.hasOwnProperty.call(e, t[c]) || !E(n[t[c]], e[t[c]]))
      return !1;
  return !0;
}
function un(n, e) {
  if (n.byteLength !== e.byteLength)
    return !1;
  const t = new Uint8Array(n), s = new Uint8Array(e);
  return V(t, s);
}
function V(n, e) {
  if (n.length !== e.length) return !1;
  for (let t = 0; t < n.length; t++)
    if (!O(n[t], e[t])) return !1;
  return !0;
}
function Zn(n, e, t) {
  const s = [], o = [], c = Object.keys(n);
  for (const i of c)
    s.push(e ? e(n[i], i, n) : i), o.push(t ? t(n[i], i, n) : n[i]);
  return [s, o];
}
function Hn(n, e) {
  const t = [], s = Object.keys(n);
  for (const o of s)
    t.push(e ? e(n[o], o, n) : [o, n[o]]);
  return t;
}
function Jn(n, e) {
  const t = [], s = Object.keys(n);
  for (const o of s)
    t.push(e ? e(n[o], o, n) : { [o]: n[o] });
  return t;
}
function Gn(...n) {
  return n.reduce((e, t) => {
    if (N(t))
      return e;
    const s = Object.keys(t);
    for (const o of s) {
      const c = e[o], i = t[o];
      (N(c) || !N(i)) && (e[o] = i);
    }
    return e;
  }, {});
}
function Yn(n) {
  const e = {};
  return Z(n, e), e;
}
function Z(n, e, t = "") {
  const s = Object.keys(n);
  for (const o of s) {
    const c = n[o], i = t ? $(n) && v(o) ? `${t}[${o}]` : `${t}.${o}` : $(n) && v(o) ? `[${o}]` : o;
    d(c) ? Z(c, e, i) : e[i] = c;
  }
}
function Qn(n, e, t) {
  const s = /* @__PURE__ */ new Map(), o = n.length;
  for (let r = 0; r < o; r++) {
    const l = t !== void 0 ? p(t) ? f(n[r], t) : t(n[r], r, n) : n[r];
    s.has(l) || s.set(l, { value: n[r], count: 1 });
  }
  const c = e.length;
  for (let r = 0; r < c; r++) {
    const l = t !== void 0 ? p(t) ? f(e[r], t) : t(e[r], r, e) : e[r], u = s.get(l);
    u && u.count++;
  }
  const i = [];
  for (const r of s.values())
    r.count === 1 && i.push(r.value);
  return i;
}
function Xn(n, e, t) {
  const s = /* @__PURE__ */ new Map(), o = n.length;
  for (let r = 0; r < o; r++) {
    const l = t !== void 0 ? p(t) ? f(n[r], t) : t(n[r], r, n) : n[r], u = s.get(l);
    u ? u.count += 1 : s.set(l, { value: n[r], count: 1 });
  }
  const c = e.length;
  for (let r = 0; r < c; r++) {
    const l = t !== void 0 ? p(t) ? f(e[r], t) : t(e[r], r, e) : e[r], u = s.get(l);
    u ? u.count += 1 : s.set(l, { value: e[r], count: 1 });
  }
  const i = [];
  for (const r of s.values())
    r.count > 1 && i.push(r.value);
  return i;
}
function gn(n, e, t) {
  const s = /* @__PURE__ */ new Map(), o = n.length;
  for (let i = 0; i < o; i++) {
    const r = t !== void 0 ? p(t) ? f(n[i], t) : t(n[i], i, n) : n[i];
    s.has(r) || s.set(r, n[i]);
  }
  const c = e.length;
  for (let i = 0; i < c; i++) {
    const r = t !== void 0 ? p(t) ? f(e[i], t) : t(e[i], i, e) : e[i];
    s.has(r) || s.set(r, e[i]);
  }
  return [...s.values()];
}
function ne(n, e, t) {
  const s = /* @__PURE__ */ new Map(), o = n.length;
  for (let r = 0; r < o; r++) {
    const l = t !== void 0 ? p(t) ? f(n[r], t) : t(n[r], r, n) : n[r];
    s.get(l) || s.set(l, { value: n[r], mask: 1 });
  }
  const c = e.length;
  for (let r = 0; r < c; r++) {
    const l = t !== void 0 ? p(t) ? f(e[r], t) : t(e[r], r, e) : e[r], u = s.get(l);
    u ? u.mask |= 2 : s.set(l, { value: e[r], mask: 2 });
  }
  const i = [];
  for (const r of s.values())
    r.mask != 3 && i.push(r.value);
  return i;
}
function ee(n, e) {
  const t = /* @__PURE__ */ new Map(), s = n.length;
  for (let c = 0; c < s; c++) {
    const i = e !== void 0 ? p(e) ? f(n[c], e) : e(n[c], c, n) : n[c];
    t.get(i) || t.set(i, n[c]);
  }
  const o = [];
  for (const c of t.values())
    o.push(c);
  return o;
}
function te(n, e, t, s, o) {
  const c = /* @__PURE__ */ new Map(), i = e.length;
  for (let u = 0; u < i; u++)
    c.set(
      p(s) ? f(e[u], s) : s(e[u], u, e),
      e[u]
    );
  const r = [], l = n.length;
  for (let u = 0; u < l; u++) {
    const a = p(t) ? f(n[u], t) : t(n[u], u, n), y = c.get(a);
    r.push(o(n[u], y));
  }
  return r;
}
const ln = (n, e) => n.localeCompare(e), fn = (n, e) => e.localeCompare(n), H = (n, e) => n - e, an = (n, e) => e - n, yn = (n, e) => n < e ? -1 : e > n ? 1 : 0, oe = (n, e) => n < e ? 1 : e > n ? -1 : 0;
function pn(n, e) {
  const t = n.map((s, o) => o);
  return t.sort((s, o) => e ? e(n[s], n[o]) : yn(`${n[s]}`, `${n[o]}`)), t;
}
function se(n, e, t) {
  const s = {}, o = n.length;
  for (let c = 0; c < o; c++) {
    const i = Object.keys(n[c])[0], r = e ? p(e) ? f(n[c], e) : e(n[c], c, n) : i, l = t ? p(t) ? f(n[c], t) : t(n[c], c, n) : n[c][i];
    s[r] = l;
  }
  return s;
}
function ce(n, e, t) {
  const s = {}, o = n.length;
  for (let c = 0; c < o; c++) {
    const i = e ? p(e) ? f(n[c], e) : e(n[c], c, n) : n[c][0], r = t ? p(t) ? f(n[c], t) : t(n[c], c, n) : n[c][1];
    s[i] = r;
  }
  return s;
}
function ie(n, e, t, s) {
  const o = {}, c = n.length;
  for (let i = 0; i < c; i++) {
    const r = t ? p(t) ? f(n[i], t) : t(n[i], i, n) : n[i], l = s ? p(s) ? f(e[i], s) : s(e[i], i, e) : e[i];
    o[r] = l;
  }
  return o;
}
function re(n, e) {
  const t = pn(n, e), s = t.map((c) => n[c]), o = n.length;
  for (let c = 0; c < o; c++)
    n[c] = s[c];
  return t;
}
function ue(n, e = "asc", t) {
  return n.sort((s, o) => {
    const c = t ? m(t) ? t(s) : f(s, t) : s, i = t ? m(t) ? t(o) : f(o, t) : o;
    return e === "asc" ? ln(c, i) : fn(c, i);
  });
}
function le(n, e, t, s = "children") {
  const o = /* @__PURE__ */ new Map(), c = n.length, i = new Array(c).fill(0), r = /* @__PURE__ */ new Map();
  for (let u = 0; u < c; u++) {
    const a = e ? m(e) ? e(n[u], u, n) : f(n[u], e) : n[u].key;
    o.set(a, n[u]);
    let y = f(n[u], s);
    $(y) || (y = [], on(n[u], s, y)), r.set(n[u], y);
  }
  for (let u = 0; u < c; u++) {
    const a = t ? m(t) ? t(n[u], u, n) : f(n[u], t) : n[u].parent, y = o.get(a);
    if (y) {
      const h = r.get(y);
      h && (h.push(n[u]), i[u]++);
    }
  }
  const l = [];
  for (let u = 0; u < c; u++)
    i[u] === 0 && l.push(n[u]);
  return l;
}
function fe(n, e) {
  if (e < 1)
    throw TypeError("Length of chunk cannot be less than 1.");
  const t = Math.floor(e), s = n.length, o = [];
  for (let c = 0; c < s; c++) {
    const i = Math.floor(c / t);
    o[i] === void 0 && (o[i] = []), o[i].push(n[c]);
  }
  return o;
}
function ae(n, e = "asc", t) {
  return n.sort((s, o) => {
    const c = t ? m(t) ? t(s) : f(s, t) : s, i = t ? m(t) ? t(o) : f(o, t) : o;
    return e === "asc" ? H(c, i) : an(c, i);
  });
}
function ye(n, e, t) {
  const s = /* @__PURE__ */ new Map();
  e.forEach((i, r) => {
    s.set(i, r);
  });
  const o = /* @__PURE__ */ new Map();
  return n.forEach((i, r) => {
    o.set(i, r);
  }), n.sort((i, r) => {
    const l = t !== void 0 ? p(t) ? f(i, t) : t(i) : i, u = t !== void 0 ? p(t) ? f(r, t) : t(r) : r, a = s.get(l), y = s.get(u);
    return a !== void 0 && y !== void 0 ? a - y : a !== void 0 ? -1 : y !== void 0 ? 1 : o.get(i) - o.get(r);
  });
}
function pe(n, e, t = 1) {
  if (n === e)
    return [];
  if (t === 0)
    throw TypeError("range step must be not equal 0.");
  const s = [];
  for (; t > 0 ? n < e : e < n; )
    s.push(n), n += t;
  return s;
}
function he(n, e = 100, t = 2) {
  if (e === 0)
    throw new TypeError("Total cannot be zero.");
  return `${(n / e * 100).toFixed(t)}%`;
}
function me(n) {
  const e = Math.abs(n).toString(), [t, s] = e.split("."), o = t.replace(/\B(?=(\d{3})+(?!\d))/g, ",");
  let c;
  if (s) {
    const i = s.replace(/(\d{3})(?=\d)/g, "$1,");
    c = `${o}.${i}`;
  } else
    c = o;
  return n < 0 ? "-" + c : c;
}
const P = ["bit", "B", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB"];
function de(n, e, t, s = "binary") {
  let o = -1, c = -1;
  const i = P.length;
  for (let y = 0; y < i && (e === P[y] && (o = y), t === P[y] && (c = y), !(o > -1 && c > -1)); y++)
    ;
  if (o === -1)
    throw TypeError("Unknown unit of from.");
  if (c === -1)
    throw TypeError("Unknown unit of to.");
  const r = s === "decimal" ? 1e3 : 1024, l = o === 0 && c > 0, u = c === 0 && o > 0, a = o - c + (l ? 1 : 0) - (u ? 1 : 0);
  return n * Math.pow(r, a) / (l ? 8 : 1) * (u ? 8 : 1);
}
function hn(n, e) {
  let t = 0;
  const s = n.length;
  for (let o = 0; o < s; o++)
    t += e !== void 0 ? p(e) ? f(n[o], e) : e(n[o], o, n) : n[o];
  return t;
}
function ve(n, e) {
  if (n.length === 0)
    throw TypeError("Length of data to calculate average must > 0");
  return hn(n, e) / n.length;
}
function mn(n, e) {
  const t = /* @__PURE__ */ new Map(), s = n.length;
  for (let o = 0; o < s; o++) {
    const c = e !== void 0 ? p(e) ? f(n[o], e) : e(n[o], o, n) : n[o];
    t.set(c, (t.get(c) ?? 0) + 1);
  }
  return t;
}
function we(n, e) {
  const t = n.map((o, c) => e !== void 0 ? p(e) ? f(o, e) : e(o, c, n) : o);
  t.sort(H);
  const s = Math.floor(t.length / 2);
  return t.length & 1 ? t[s] : (t[s] + t[s - 1]) / 2;
}
function ke(n, e) {
  const t = mn(n, e), s = Math.max(...t.values());
  let o;
  for (const c of t.entries())
    c[1] === s && (o = c[0]);
  return o;
}
function Ae(n, e) {
  let t = -1 / 0;
  const s = n.length;
  for (let o = 0; o < s; o++) {
    const c = e !== void 0 ? p(e) ? f(n[o], e) : e(n[o], o, n) : n[o];
    c > t && (t = c);
  }
  return t;
}
function be(n, e) {
  let t = -1 / 0, s = n[0];
  const o = n.length;
  for (let c = 0; c < o; c++) {
    const i = e !== void 0 ? p(e) ? f(n[c], e) : e(n[c], c, n) : n[c];
    i > t && (t = i, s = n[c]);
  }
  return s;
}
function xe(n, e) {
  let t = 1 / 0;
  const s = n.length;
  for (let o = 0; o < s; o++) {
    const c = e !== void 0 ? p(e) ? f(n[o], e) : e(n[o], o, n) : n[o];
    c < t && (t = c);
  }
  return t;
}
function Oe(n, e) {
  let t = 1 / 0, s = n[0];
  const o = n.length;
  for (let c = 0; c < o; c++) {
    const i = e !== void 0 ? p(e) ? f(n[c], e) : e(n[c], c, n) : n[c];
    i < t && (t = i, s = n[c]);
  }
  return s;
}
function Me(n, e) {
  const t = /* @__PURE__ */ new Map();
  let s = -1 / 0, o = 0;
  const c = n.length;
  for (let i = 0; i < c; i++) {
    const r = e !== void 0 ? p(e) ? f(n[i], e) : e(n[i], i, n) : n[i];
    let l = t.get(r);
    l || (l = { count: 0, firstIndex: i }, t.set(r, l)), l.count++, l.count > s && (s = l.count, o = l.firstIndex);
  }
  return n[o];
}
function Te() {
  return Math.random() < 0.5;
}
function J(n = 0, e = 1) {
  return n + Math.random() * (e - n);
}
function Ie(n) {
  const e = Math.floor(J(0, n.length));
  return n[e];
}
const dn = "abcdefghijklmnopqrstuvwxyz", vn = "ABCDEFGHIJKLMNOPQRSTUVWXYZ", wn = "0123456789", kn = "!@#$%^&*()_+-=[]{}|;:,.<>?";
function Se(n, e = {}) {
  if (n === 0)
    return "";
  let t = "";
  const {
    lowercase: s = !0,
    uppercase: o = !0,
    number: c = !0,
    symbol: i = !1,
    customized: r
  } = e;
  if (s && (t += dn), o && (t += vn), c && (t += wn), i && (t += kn), r && (t += r), t.length === 0)
    return "";
  const l = t.length;
  let u = "";
  for (let a = 0; a < n; a++)
    u += t.charAt(Math.floor(J(0, l)));
  return u;
}
function Ee(n = 0, e = 10) {
  return n + Math.floor(Math.random() * (e - n));
}
function Be(n, e = "right") {
  return function(...s) {
    let o, c;
    return e === "left" ? (o = n[0](...s), c = n.slice(1)) : (o = n[n.length - 1](...s), c = n.slice(0, n.length - 1)), e === "left" ? c.reduce((i, r) => r(i), o) : c.reduceRight((i, r) => r(i), o);
  };
}
function Ce(n) {
  const e = n.length;
  return function t(...s) {
    return s.length >= e ? n.apply(this, s) : function(...o) {
      return t.apply(this, s.concat(o));
    };
  };
}
function Ne(n, e, t) {
  let s, o;
  const { immediate: c = !1, maxWait: i } = t || {};
  return function(...l) {
    i !== void 0 && !o && (o = setTimeout(() => {
      o = void 0, n.apply(this, l);
    }, i)), !s && c && n.apply(this, l), s && clearTimeout(s), s = setTimeout(() => {
      o && (clearTimeout(o), o = void 0), n.apply(this, l);
    }, e);
  };
}
function Pe(n, e) {
  return function(...t) {
    setTimeout(() => {
      n.apply(this, t);
    }, e);
  };
}
const An = (...n) => JSON.stringify(n);
function $e(n, e, t) {
  const s = /* @__PURE__ */ new Map(), o = function(...c) {
    const i = e ? e(...c) : An(...c);
    if (s.has(i))
      return s.get(i);
    const r = n.apply(this, c);
    return s.set(i, r), r;
  };
  return t && o(...t), o;
}
function Fe(n, e, t) {
  let s = 0, o;
  const { leading: c = !1, trailing: i = !0 } = t || {};
  return function(...l) {
    const u = +/* @__PURE__ */ new Date();
    c === !1 && (s = u);
    const a = e - (u - s);
    a <= 0 && (o && (clearTimeout(o), o = void 0), s = u, n.apply(this, l)), !o && i && (o = setTimeout(() => {
      o = void 0, s = c ? +/* @__PURE__ */ new Date() : 0, n.apply(this, l);
    }, a));
  };
}
function Ue(n) {
  let e = !1, t;
  return function(...s) {
    return e || (e = !0, t = n.apply(this, s)), t;
  };
}
function je(n, e) {
  return function(...t) {
    return n.apply(this, t) ?? e;
  };
}
async function _e(n) {
  try {
    return [await n(), null];
  } catch (e) {
    return [null, e];
  }
}
async function qe(n, e, t = "concurrent") {
  let s = n.length;
  if (!s)
    return;
  let o;
  const c = new Promise((i) => {
    o = i;
  });
  for (const [i, r] of n.entries())
    t === "sequential" ? await e(r, i, n) : Promise.resolve(e(r, i, n)).finally(() => {
      s--, s || o();
    });
  t === "concurrent" && await c;
}
async function Le(n, e, t = "concurrent") {
  const s = Object.keys(n);
  let o = s.length;
  if (!o)
    return;
  let c;
  const i = new Promise((r) => {
    c = r;
  });
  for (const r of s)
    t === "sequential" ? await e(n[r], r, n) : Promise.resolve(e(n[r], r, n)).finally(() => {
      o--, o || c();
    });
  t === "concurrent" && await i;
}
async function De(n, e, t = "concurrent") {
  let s = n.length;
  if (!s)
    return [];
  let o;
  const c = new Promise((r) => {
    o = r;
  }), i = [];
  for (const [r, l] of n.entries())
    t === "sequential" ? i[r] = await e(l, r, n) : Promise.resolve(await e(l, r, n)).then((u) => {
      i[r] = u;
    }).finally(() => {
      s--, s || o();
    });
  return t === "concurrent" && await c, i;
}
async function Re(n, e) {
  const t = [], s = [];
  for (const o of n) {
    const c = Promise.resolve().then(() => o());
    if (t.push(c), e <= t.length) {
      const i = c.finally(() => {
        s.splice(s.indexOf(i), 1);
      });
      s.push(i), s.length >= e && await Promise.race(s);
    }
  }
  return Promise.allSettled(t);
}
async function ze(n, e) {
  const t = [], s = en(e) ? void 0 : { status: "fulfilled", value: e };
  for (const [o, c] of n.entries())
    t[o] = await Promise.resolve().then(
      () => Promise.resolve(c(o > 0 ? t[o - 1] : s))
    ).then((i) => ({ status: "fulfilled", value: i })).catch((i) => ({ status: "rejected", reason: i }));
  return t;
}
const bn = (
  // @ts-ignore
  typeof process < "u" && process.env.NODE_ENV !== "production"
), G = "Parsnip-Kit";
function L(...n) {
  bn && console.warn(`[${G}]`, ...n);
}
function D(n) {
  return new Error(`[${G}] ${n}`);
}
function We(n, e = 3, t) {
  const {
    delay: s = 300,
    delayFactor: o = 2,
    shouldRetry: c = (u, a) => !0,
    // eslint-disable-line
    onSuccess: i = () => {
    },
    onFailure: r = () => {
    }
  } = t || {};
  async function l(...u) {
    let a = 0;
    const y = async () => {
      try {
        const h = await n.apply(this, u);
        return i(h, a + 1), {
          value: h,
          status: "fulfilled"
        };
      } catch (h) {
        if (a += 1, a > e)
          return r(h, a), {
            reason: [h, D("Max retries exceeded")],
            status: "rejected"
          };
        if (c(h, a)) {
          r(h, a);
          const b = s * Math.pow(o, a);
          return await new Promise((w) => setTimeout(w, b)), y();
        } else
          return r(h, a), {
            reason: [
              h,
              D("Retry canceled by options.shouldRetry")
            ],
            status: "rejected"
          };
      }
    };
    return y();
  }
  return function(...u) {
    return l.apply(this, u);
  };
}
function Ke(n, e, t) {
  const {
    maxRetries: s = 3,
    maxCalls: o = 1 / 0,
    sequential: c = !1,
    leading: i = !0,
    onSuccess: r = () => {
    },
    onFailure: l = () => {
    }
  } = t || {};
  function u(...a) {
    let y, h = 0, b = 0, w = !1;
    const T = async () => {
      b++;
      try {
        const I = await n.apply(this, a);
        r(I, h + 1), h = 0;
      } catch (I) {
        h++, l(I, h), h > s && (C(), L(`Max retries exceeded: ${s}. Last error:`, I));
      }
      b >= o && (C(), L(`Max calls exceeded: ${o}. Stopping polling.`));
    }, B = async () => {
      y = setTimeout(async () => {
        w || (c ? (await T(), B()) : (B(), T()));
      }, e);
    }, F = async () => {
      i && (c ? await T() : T()), B();
    }, C = () => {
      w = !0, y && (clearTimeout(y), y = void 0);
    };
    return F(), {
      stop: C,
      isRunning: () => !w,
      start: () => {
        w && (h = 0, b = 0, w = !1, F());
      }
    };
  }
  return function(...a) {
    return u.apply(this, a);
  };
}
export {
  qe as asyncForEach,
  Le as asyncForEachFields,
  De as asyncMap,
  ve as average,
  xn as camelCase,
  X as capitalize,
  fe as chunk,
  Dn as clone,
  Rn as cloneDeep,
  yn as codeUnitComparatorAsc,
  oe as codeUnitComparatorDesc,
  Be as combine,
  Re as concurrent,
  de as convertDataUnit,
  mn as count,
  Ce as curry,
  Ne as debounce,
  Pe as delay,
  Ln as deleteByPath,
  Qn as difference,
  U as escapeRegExp,
  Kn as filterFields,
  Yn as flattenObject,
  Vn as forEachFields,
  f as getByPath,
  S as getTypeTag,
  _e as go,
  Bn as htmlDecode,
  En as htmlEncode,
  Xn as intersection,
  $ as isArray,
  Fn as isBigInt,
  Nn as isBoolean,
  Un as isClass,
  E as isEqual,
  O as isEqualStrict,
  qn as isFloat,
  m as isFunction,
  R as isInfinity,
  _n as isInt,
  z as isNanValue,
  Pn as isNull,
  N as isNullish,
  k as isNumber,
  v as isNumberString,
  d as isObject,
  tn as isObjectLike,
  j as isPrimitive,
  jn as isPseudoArray,
  p as isString,
  $n as isSymbol,
  en as isUndefined,
  se as joinToObject,
  On as kebabCase,
  te as leftJoin,
  ue as lexSort,
  le as linkToTree,
  Ae as max,
  be as maxItem,
  we as median,
  $e as memoize,
  Gn as mergeSkipNullish,
  xe as min,
  Oe as minItem,
  ke as mode,
  Me as modeItem,
  H as numberComparatorAsc,
  an as numberComparatorDesc,
  ae as numberSort,
  Hn as objectToPairs,
  zn as omit,
  Ue as once,
  ye as orderSort,
  ce as pairsToObject,
  Cn as parseTemplate,
  Mn as pascalCase,
  he as percent,
  Wn as pick,
  Ke as poll,
  Te as randomBoolean,
  Ie as randomFromArray,
  Ee as randomInt,
  J as randomNumber,
  Se as randomString,
  pe as range,
  We as retry,
  ze as sequential,
  on as setByPath,
  Tn as snakeCase,
  pn as sortIndex,
  re as sortWithIndex,
  Jn as splitToArrays,
  M as splitToKeys,
  A as splitToWords,
  ln as stringComparatorAsc,
  fn as stringComparatorDesc,
  hn as sum,
  ne as symmetricDifference,
  me as thousandSeparator,
  Fe as throttle,
  Sn as titleCase,
  gn as union,
  ee as unique,
  Zn as unzipToArrays,
  In as upperSnakeCase,
  je as withFallback,
  ie as zipToObject
};
