// @pixelium/web-vue/es/link/index.js
import './css.js'
import { defineComponent, useSlots, computed, ref, createElementBlock, openBlock, normalizeStyle, normalizeClass, createCommentVNode, renderSlot, unref, createBlock } from "vue";
import { p as parseColor, a as generatePalette, r as rgbaColor2string } from "../share/util/color.js";
import { S as SpinnerThirdSolid } from "../auto-complete/index.js";
import { u as useDarkMode } from "../share/hook/use-dark-mode.js";
const _hoisted_1 = ["href", "target"];
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "Link"
  },
  __name: "index",
  props: {
    disabled: { type: Boolean },
    loading: { type: Boolean },
    theme: null,
    color: null,
    href: null,
    target: null
  },
  setup(__props) {
    const props = __props;
    const slots = useSlots();
    const darkMode = useDarkMode();
    const palette = computed(() => {
      if (!props.color) return null;
      const color = parseColor(props.color);
      const palette2 = generatePalette(color.r, color.g, color.b, color.a, darkMode.value);
      return palette2;
    });
    const hoverFlag = ref(false);
    const activeFlag = ref(false);
    const toggleHover = (value) => {
      hoverFlag.value = value;
    };
    const toggleActive = (value) => {
      activeFlag.value = value;
    };
    const getTextColorWithPalette = (palette2, disabled, loading, hoverFlag2, activeFlag2) => {
      if (!palette2) return void 0;
      if (disabled) {
        return rgbaColor2string(palette2[1]);
      }
      if (loading) return rgbaColor2string(palette2[5]);
      if (activeFlag2) return rgbaColor2string(palette2[6]);
      if (hoverFlag2) return rgbaColor2string(palette2[4]);
      return rgbaColor2string(palette2[5]);
    };
    const textColor = computed(() => {
      return getTextColorWithPalette(
        palette.value,
        props.disabled,
        props.loading,
        hoverFlag.value,
        activeFlag.value
      );
    });
    const clickHandler = (e) => {
      if (props.disabled || props.loading) {
        e.preventDefault();
      }
    };
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("a", {
        class: normalizeClass(["pixelium px-link", {
          "px-link__loading": props.loading,
          "px-link__disabled": props.disabled,
          "px-link__custom": palette.value,
          [`px-link__${props.theme || "primary"}`]: true
        }]),
        style: normalizeStyle({
          color: textColor.value
        }),
        onMouseenter: _cache[0] || (_cache[0] = ($event) => toggleHover(true)),
        onMouseleave: _cache[1] || (_cache[1] = ($event) => toggleHover(false)),
        onMousedown: _cache[2] || (_cache[2] = ($event) => toggleActive(true)),
        onMouseup: _cache[3] || (_cache[3] = ($event) => toggleActive(false)),
        href: props.href,
        target: props.target,
        onClick: clickHandler
      }, [
        unref(slots).icon || props.loading ? (openBlock(), createElementBlock(
          "div",
          {
            key: 0,
            class: normalizeClass(["px-link-icon-wrapper", {
              "px-link-icon-wrapper__last": !unref(slots).default
            }])
          },
          [
            props.loading ? (openBlock(), createBlock(unref(SpinnerThirdSolid), {
              key: 0,
              class: "px-link-icon px-animation__loading",
              style: normalizeStyle({
                fill: textColor.value
              })
            }, null, 8, ["style"])) : renderSlot(_ctx.$slots, "icon", { key: 1 })
          ],
          2
          /* CLASS */
        )) : createCommentVNode("v-if", true),
        renderSlot(_ctx.$slots, "default")
      ], 46, _hoisted_1);
    };
  }
});
export {
  _sfc_main as _
};
