// @pixelium/web-vue/es/message-box/message-box-wrapped.js
import { defineComponent, useAttrs, ref, createBlock, openBlock, Teleport, createVNode, mergeProps, unref } from "vue";
import { _ as _sfc_main$1 } from "./message-box.js";
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "MessageBox"
  },
  __name: "message-box-wrapped",
  props: {
    messages: null,
    placement: null,
    position: null,
    zIndex: null,
    root: { default: "body" },
    "onUpdate:messages": { type: Function },
    onClose: { type: Function }
  },
  setup(__props, { expose: __expose }) {
    const props = __props;
    const attrs = useAttrs();
    const messageBoxRef = ref(null);
    __expose({
      close: (id) => {
        var _a;
        (_a = messageBoxRef.value) == null ? void 0 : _a.close(id);
      }
    });
    return (_ctx, _cache) => {
      return openBlock(), createBlock(Teleport, {
        to: props.root
      }, [
        createVNode(
          _sfc_main$1,
          mergeProps({ ...props, ...unref(attrs) }, {
            ref_key: "messageBoxRef",
            ref: messageBoxRef
          }),
          null,
          16
          /* FULL_PROPS */
        )
      ], 8, ["to"]);
    };
  }
});
export {
  _sfc_main as _
};
