// @pixelium/web-vue/es/message-box/message-box.js
import './css.js'
import { defineComponent, useModel, computed, shallowRef, createElementBlock, openBlock, normalizeStyle, unref, createElementVNode, normalizeClass, Fragment, renderList, createBlock } from "vue";
import { N } from "../aside/index.js";
import { _ as _sfc_main$1 } from "../message/index.js";
import { n as nanoid } from "./index.js";
import { u as useZIndex } from "../share/hook/use-z-index.js";
const _sfc_main = /* @__PURE__ */ defineComponent({
  ...{
    name: "MessageBoxInner"
  },
  __name: "message-box",
  props: {
    messages: {},
    placement: { default: "top" },
    position: { default: "top" },
    zIndex: {},
    root: {},
    "onUpdate:messages": {}
  },
  emits: ["close"],
  setup(__props, { expose: __expose, emit: __emit }) {
    const props = __props;
    const [currentZIndex] = useZIndex("message");
    const messages = useModel(props, "messages");
    const emits = __emit;
    const clone = computed(() => {
      return messages.value.map((e) => {
        return {
          ...e,
          id: e.id ?? nanoid()
        };
      });
    });
    const closeHandler = (id) => {
      if (N(id)) {
        return;
      }
      if (props["onUpdate:messages"]) {
        const idx = clone.value.findIndex((item) => item.id === id);
        if (idx >= 0) {
          const cur = [...clone.value];
          cur.splice(idx, 1);
          messages.value = cur;
        }
      }
      emits("close", id);
    };
    const messageRef = shallowRef([]);
    __expose({
      close: (id) => {
        var _a;
        const idx = clone.value.findIndex((e) => e.id === id);
        if (idx === -1) {
          return;
        }
        (_a = messageRef.value[idx]) == null ? void 0 : _a.close();
      }
    });
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock(
        "div",
        {
          class: "px-message-box",
          style: normalizeStyle({
            zIndex: props.zIndex ?? unref(currentZIndex)
          })
        },
        [
          createElementVNode(
            "div",
            {
              class: normalizeClass(["px-message-box-container", {
                [`px-message-box-container__${props.placement || props.position || "top"}`]: true
              }])
            },
            [
              (openBlock(true), createElementBlock(
                Fragment,
                null,
                renderList(clone.value, (message) => {
                  return openBlock(), createBlock(_sfc_main$1, {
                    key: message.id,
                    duration: message.duration,
                    content: message.content,
                    id: message.id,
                    onClose: closeHandler,
                    type: message.type,
                    color: message.color,
                    icon: message.icon,
                    closable: message.closable,
                    ref_for: true,
                    ref_key: "messageRef",
                    ref: messageRef
                  }, null, 8, ["duration", "content", "id", "type", "color", "icon", "closable"]);
                }),
                128
                /* KEYED_FRAGMENT */
              ))
            ],
            2
            /* CLASS */
          )
        ],
        4
        /* STYLE */
      );
    };
  }
});
export {
  _sfc_main as _
};
